rules/store.go

```diff
@@ -0,0 +1,191 @@
+
+	// ReadRuleSet returns the version and the persisted ruleset data in kv store.
+	ReadRuleSet(namespaceName string) (MutableRuleSet, error)
+
+	// RuleSetKey generates a ruleset key for a given namespace
+	ruleSetKey(namespaceName string) string
+}
+
+// StoreOptions is a configuration for the rules r/w store.
+type StoreOptions struct {
+	NamespacesKey string
+	RuleSetKeyFmt string
+}
+
+type store struct {
+	kvStore kv.TxnStore
+	opts    StoreOptions
+}
+
+// NewStoreOptions creates a new store options struct.
+func NewStoreOptions(namespacesKey string, rulesetKeyFmt string) StoreOptions {
+	return StoreOptions{NamespacesKey: namespacesKey, RuleSetKeyFmt: rulesetKeyFmt}
+}
+
+// NewStore creates a new Store.
+func NewStore(kvStore kv.TxnStore, opts StoreOptions) Store {
+	return store{kvStore: kvStore, opts: opts}
+}
+
+func (s store) ReadNamespaces() (*Namespaces, error) {
+	value, err := s.kvStore.Get(s.opts.NamespacesKey)
+	if err != nil {
+		return nil, err
+	}
+
+	version := value.Version()
+	var namespaces schema.Namespaces
+	if err := value.Unmarshal(&namespaces); err != nil {
+		return nil, err
+	}
+
+	nss, err := NewNamespaces(version, &namespaces)
+	if err != nil {
+		return nil, err
+	}
+	return &nss, err
+}
+
+func (s store) ruleSetKey(ns string) string {
+	return fmt.Sprintf(s.opts.RuleSetKeyFmt, ns)
+}
+
+func (s store) ReadRuleSet(nsName string) (MutableRuleSet, error) {
+	ruleSetKey := s.ruleSetKey(nsName)
+	value, err := s.kvStore.Get(ruleSetKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	version := value.Version()
+	var ruleSet schema.RuleSet
+	if err := value.Unmarshal(&ruleSet); err != nil {
+		return nil, fmt.Errorf("Could not fetch RuleSet %s: %v", nsName, err.Error())
+	}
+	rs, err := NewMutableRuleSetFromSchema(version, &ruleSet)
+	if err != nil {
+		return nil, fmt.Errorf("Could not fetch RuleSet %s: %v", nsName, err.Error())
+	}
+	return rs, err
+}
+
+func (s store) ruleSetTransaction(rs MutableRuleSet) (kv.Condition, kv.Op, error) {
+	if rs == nil {
+		return nil, nil, errNilRuleSet
+	}
+
+	ruleSetKey := s.ruleSetKey(string(rs.Namespace()))
+	rsSchema, err := rs.Schema()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ruleSetCond := kv.NewCondition().
+		SetKey(ruleSetKey).
+		SetCompareType(kv.CompareEqual).
+		SetTargetType(kv.TargetVersion).
+		SetValue(rs.Version())
+
+	return ruleSetCond, kv.NewSetOp(ruleSetKey, rsSchema), nil
+}
+
+func (s store) namespacesTransaction(nss *Namespaces) (kv.Condition, kv.Op, error) {
+	if nss == nil {
+		return nil, nil, errNilNamespaces
+	}
+
+	namespacesKey := s.opts.NamespacesKey
+	nssSchema, err := nss.Schema()
+	if err != nil {
+		return nil, nil, err
+	}
+	namespacesCond := kv.NewCondition().
+		SetKey(namespacesKey).
+		SetCompareType(kv.CompareEqual).
+		SetTargetType(kv.TargetVersion).
+		SetValue(nss.Version())
+
+	return namespacesCond, kv.NewSetOp(namespacesKey, nssSchema), nil
+}
+
+func (s store) WriteRuleSet(rs MutableRuleSet) error {
```

Comment:
nit: my personal preference is to put the exported methods at the top of the file and the unexported ones at the bottom so users only need to pay attention to the top half if they only care about APIs.